File: algo/CountingSort/countingSort_advance.py
class countingSort:
    @classmethod
    def countingSort_advance(cls, arr):
        if len(arr) == 0 or len(arr) == 1:
            return arr
        max = arr[0]
        min = arr[0]
        for i in range(len(arr)):
            if arr[i] > max:
                max = arr[i]
            elif arr[i] < min:
                min = arr[i]
        total_range = max - min + 1

        counting = [0] * total_range
        for element in arr:
            idx = element - min
            counting[idx] += 1
        print(counting)

        precount = 0
        for i in range(0, total): #why total??
            #注意, 自己想不出來
            precount += counting[i]
            counting[i] = precount - counting[i]
        print(counting)
        # To Figure
        result = [0] * len(arr)
        for element in arr:
            position = counting[element - min] #why min??
            result[position] = element
            counting[element - min] += 1

        for i in range(len(arr)):
            arr[i] = result[i]

if __name__ == "__main__":
    arr = [8, 7, 1, 2, 8, 6, 8, 2]
    countingSort.countingSort_advance(arr)
    print(arr)
"""
代碼出處:
https://leetcode.cn/leetbook/read/sort-algorithms/ozyo63/
"""

File: algo/CountingSort/CountingSort_advance.java
import java.util.Arrays;
public class CountingSort_advance {
    void CountingSort(int[] arr){
        int max = arr[0];
        int min = arr[0];
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] > max)
                max = arr[i];
            else if ((arr[i] < min))
                min = arr[i];
        }
        int total_range = max - min + 1;
        int[] counting = new int[total_range];
        for (int element: arr){
            counting[element - min] ++;
        }
        System.out.println(Arrays.toString(counting));
        int preSum = 0;
        for (int i = 0; i < total_range; i++) {
            int temp = counting[i];
            counting[i] = preSum;
            preSum += temp;
        }
        System.out.println(Arrays.toString(counting));
        int[] result = new int[arr.length];
        for (int element: arr){
            int position = counting[element - min];//element - min 才能映射出正確位置, 可以從上個迴圈推得
            result[position] = element;
            counting[element - min]++;
        }
        System.arraycopy(result, 0,  arr, 0,  result.length);
    }
    public static void main(String[] args) {
        int[] arr = {2, 4, 5, 2, 6, 4};
        CountingSort_advance Sort = new CountingSort_advance();
        Sort.CountingSort(arr);
        System.out.println(Arrays.toString(arr));
    }
}
